### PR TITLE
Feat(silu_mul): ascend compiler with 2 patterns

### PR DESCRIPTION
**Pull request by wxue1**
1, fused silu mul pattern
	2, add rms_norm pattern
	3, fix tab error

### CODE REVIEW

File: xpu_graph/passes/patterns/targets/ascend/fuse_silu_mul.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+from .triton_kernel.fused_silu_mul import fused_silu_mul
+
+class AddSilumulOperation(nn.Module):
+    def forward(self, input):
+        return torch.ops.torch_npu_triton.fused_silu_mul(input)
+
+
+class FusedSilumul(Pattern):
+    _opt_level = OptLevel.level2
```
[the remark this answers is not in the document]
Comment (wxue1) [reply xpu_graph/passes/patterns/targets/ascend/fuse_silu_mul.py:36]: 就是silu算子和mul算子的组合

File: xpu_graph/passes/patterns/utils/check_ops.py [hunk 2]
```diff
@@ -229,6 +229,13 @@ def check_trans_op(node: fx.Node) -> bool:
 def check_t_op(node: fx.Node) -> bool:
     return check_op(node, aten.t.default)
 
+def check_typecast_op(node: fx.Node) -> bool:
+    if node.target == 'to' or "dtype_cast" in str(node.target):
+        return True, "to"
+    if "_to_copy" in str(node.target):
```
[the remark this answers is not in the document]
Comment (wxue1) [reply xpu_graph/passes/patterns/utils/check_ops.py]: > to_copy和to还可能和h2d、d2h拷贝、memory_format转换等有关联，建议做更加详细的参数检查

done

File: xpu_graph/passes/patterns/utils/check_ops.py [hunk 3]
```diff
@@ -283,6 +291,8 @@ def check_norm_op(node: fx.node):
         return True, "layer_norm"
     if node.target == "rms_norm_op":
         return True, "rms_norm"
+    if node.target == torch.ops.npu.npu_rms_norm.default:
```
Comment (yfchen-byted) [inline xpu_graph/passes/patterns/utils/check_ops.py:294]: npu代码还是放在targets/npu目录里吧

**Review by mingfa-byted — APPROVED** [review]

File: xpu_graph/passes/patterns/targets/ascend/__init__.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import pkgutil
+import importlib
+
+from xpu_graph.passes.patterns.pattern import Pattern, AutoMatchPattern, PatternGroup
+from xpu_graph.config import XpuGraphConfig
+from xpu_graph.utils import logger
+
+
+def get_all_patterns(config: XpuGraphConfig):
```
[the remark this answers is not in the document]
Comment (mingfa-byted) [reply xpu_graph/passes/patterns/targets/ascend/__init__.py:9]: https://github.com/BD-Seed-HHW/xpu_graph/pull/275

File: .github/workflows/ci-for-npu.yml [hunk 5]
```diff
@@ -63,7 +63,7 @@ jobs:
         export HTTP_PROXY=${XPU_GPRAH_PROXY} HTTPS_PROXY=${XPU_GPRAH_PROXY}
         pip install -e '.[dev]'
         echo "Running npu tests..."
-        python3 -m pytest --cov=xpu_graph --ignore=tests/common/test_inference.py tests/common tests/npu
```
Comment (LeoLau94) [inline .github/workflows/ci-for-npu.yml:66]: 为啥去掉了codecov咧？

File: xpu_graph/passes/optimizer.py [hunk 6]
```diff
@@ -45,8 +45,8 @@ def __call__(self, gm: fx.GraphModule) -> bool:
                 f"{self.__class__.__bases__[0].__name__}.{self} changed graph, before: {prev_nodes_num} nodes, after: {len(gm.graph.nodes)} nodes."
             )
 
-        if changed and self._dump_graph:
-            self.__dump_files(gm)
+        # if changed and self._dump_graph:
```
Comment (LeoLau94) [inline xpu_graph/passes/optimizer.py:48]: 这个是之前debug的时候关掉的？得打开回来吧

**Comment by LeoLau94** [comment]
没有单元测试呢，补一下哈

File: xpu_graph/acc_utils.py
Comment (LeoLau94) [inline xpu_graph/acc_utils.py:1]: 这个是不是应该放在tests下的utils里

Comment (LeoLau94) [reply xpu_graph/acc_utils.py:1]: 看了下有个`xpu_graph/test_util.py`，可以放在那里？不过我是觉得应该也挪到`tests`下

File: xpu_graph/acc_utils.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
+import numpy as np
+import torch
+import torch_npu
+
+eval_standard = {
```
Comment (LeoLau94) [inline xpu_graph/acc_utils.py:5]: 这个容忍是不是应该因算子而异，从通用的角度来说，我们是不是可以考虑把这个配置设计成类，然后由使用者提供，而不是统一定义在这里。

File: xpu_graph/acc_utils.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+
+def assert_close(gold: torch.Tensor, act: torch.Tensor, eval_type: str = "DEFAULT"):
+    gold = gold.cpu()
+    act = act.cpu()
+    if act.dtype == torch.float16 or act.dtype == torch.float32 or act.dtype == torch.bfloat16:
+        assert gold.dtype == torch.float32, "golden should be f32"
+        assert not (torch.isnan(act).any() or torch.isinf(act).any()), "actual tensor can not have 'inf' or 'nan'"
+    eps = eval_standard[act.dtype]["small_value"]
```
Comment (LeoLau94) [inline xpu_graph/acc_utils.py:33]: 所以，`small_value`为啥不直接叫做`eps` doge.jpg

File: xpu_graph/acc_utils.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+        not_close = torch.where(mask, ae > atol, re > rtol)
+        print(f"count not_close = {torch.sum(not_close).item()}")
+        print(f"not_close.numel = {not_close.numel()}, gold.numel = {gold.numel()}")
+        print(f"not close ratio = {torch.sum(not_close).item()/not_close.numel()}")
+        if not torch.any(not_close):
+            return False
+
+        assert (
+            torch.sum(not_close).item() < not_close.numel() * eps
```
Comment (LeoLau94) [inline xpu_graph/acc_utils.py:52]: 这是什么比对准则？求和小于个数乘以某个小比例？

File: xpu_graph/acc_utils.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+        "small_value_atol": 1e-5,
+        "etol": 1e-3,
+    },
+}
+
+
+def assert_close(gold: torch.Tensor, act: torch.Tensor, eval_type: str = "DEFAULT"):
+    gold = gold.cpu()
+    act = act.cpu()
+    if act.dtype == torch.float16 or act.dtype == torch.float32 or act.dtype == torch.bfloat16:
```
Comment (LeoLau94) [inline xpu_graph/acc_utils.py:30]: 试试用act.dtype in [torch.float16,...]

File: xpu_graph/acc_utils.py (hunk 9 above)
Comment (LeoLau94) [reply xpu_graph/acc_utils.py:52]: 哦哦，where输出了0-1矩阵

File: xpu_graph/acc_utils.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+you can use 'benchmark_compare_close' function to compare again!"
+    elif eval_type == "ABS":
+        act = act.to(gold.dtype)
+        assert torch.equal(gold, act), "actual tensor and golden tensor are not binary equal!"
+    else:
+        assert 0, "ERROR! invalid eval_type"
+    return False
+
+
+def benchmark_compare_close(gold: torch.Tensor, act: torch.Tensor, std: torch.tensor):
```
Comment (LeoLau94) [inline xpu_graph/acc_utils.py:63]: 我理解，xpu_graph不是必须和fp32 golden比较，我们的事情只要保证和同类型的eager结果能对齐即可。

**Comment by wxue1** [comment]
> 我想这里ut应该分两种？一种是triton kernel本身的精度守护ut，另一种是pattern rewrite的ut。放在一起守护也不是不可以，但可能要多加点case了。

对的，后续重构可以考虑一下

File: xpu_graph/acc_utils.py (hunk 11 above)
Comment (wxue1) [reply xpu_graph/acc_utils.py:63]: > 另外如果要守护kernel的话，建议是放在kernel下去测试，而且得多加点case

是的

File: xpu_graph/acc_utils.py (hunk 9 above)
Comment (wxue1) [reply xpu_graph/acc_utils.py:52]: > 但是离群值数量的容忍为啥是参考eps作为比例

后续issue跟踪

File: xpu_graph/acc_utils.py (hunk 7 above)
Comment (wxue1) [reply xpu_graph/acc_utils.py:5]: > 这个我觉得可以留待日后重构，记个issue即可

okk

Comment (wxue1) [reply xpu_graph/acc_utils.py:5]: > 这个容忍是不是应该因算子而异，从通用的角度来说，我们是不是可以考虑把这个配置设计成类，然后由使用者提供，而不是统一定义在这里。

是的，这些我记录到issue里去

File: xpu_graph/acc_utils.py (hunk 8 above)
Comment (wxue1) [reply xpu_graph/acc_utils.py:33]: > 所以，`small_value`为啥不直接叫做`eps` doge.jpg

应该改个名，表示一个波动范围

File: xpu_graph/acc_utils.py (hunk 9 above)
Comment (wxue1) [reply xpu_graph/acc_utils.py:52]: > 这是什么比对准则？索引求和 小于 个数乘以某个小比例？

后续issue跟踪

Comment (wxue1) [reply xpu_graph/acc_utils.py:52]: > 哦哦，where输出了0-1矩阵

嗯

File: xpu_graph/acc_utils.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        if not torch.any(not_close):
+            return False
+
+        assert (
+            torch.sum(not_close).item() < not_close.numel() * eps
+        ), "actual tensor are not close enough with golden tensor,\
+you can use 'benchmark_compare_close' function to compare again!"
+    elif eval_type == "ABS":
+        act = act.to(gold.dtype)
+        assert torch.equal(gold, act), "actual tensor and golden tensor are not binary equal!"
```
[the remark this answers is not in the document]
Comment (wxue1) [reply xpu_graph/acc_utils.py:57]: > 浮点不能直接用吧

对，要改的

File: xpu_graph/acc_utils.py
Comment (wxue1) [reply xpu_graph/acc_utils.py:1]: > > 挪到`tests`下
> 
> tests目录不在local的PYTHONPATH里，不利于本地自测，还是放tests_utils.py吧

那我的accu_utils.py 也还是类似tests_utils.py这个层级放这

File: xpu_graph/acc_utils.py (hunk 9 above)
Comment (wxue1) [reply xpu_graph/acc_utils.py:52]: > 目前离群的情况是什么样的

这个对精度工具的具体细则得再找内部对对

File: tests/npu/test_silu_mul.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        target=Target.npu,
+        opt_level=OptLevel.level2,
+        vendor_compiler_config={"mode": "reduce-overhead", "compiler": "ge"},
+    )
+    compiled_model = torch.compile(model_forward, backend=XpuGraph(xconf), fullgraph=True)
+    res_tri = compiled_model(input)
+
+    # accuracy check
+    try:
+        assert_close(res_goden.to(torch.float32), res_tri.to(torch.float16))
```
Comment (yfchen-byted) [inline tests/npu/test_silu_mul.py:55]: 留待后续修复：参考common目录test_fused_layernorm，通过caplog检查pass有没有生效

File: xpu_graph/passes/patterns/targets/npu/fuse_add_rmsnorm.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            nodes_to_remove = [
+                getitem_final_rmsnorm,
+                final_rmsnorm,
+                to_node,
+            ]
+            for n in nodes_to_remove:
+                graph.erase_node(n)
+
+        if changed:
+            gm.graph.lint()
```
Comment (yfchen-byted) [inline xpu_graph/passes/patterns/targets/npu/fuse_add_rmsnorm.py:116]: 不要加，pattern manager会做

File: xpu_graph/passes/patterns/targets/npu/fuse_add_rmsnorm.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+class FusedAddRmsnorm(Pattern):
+    _opt_level = OptLevel.level2
+
+    def _match_pattern(self, getitem_final_rmsnorm: Node) -> Optional[List[Node]]:
+        if not check_getitem_op(getitem_final_rmsnorm):
+            return None
+        final_rmsnorm = getitem_final_rmsnorm.args[0]
+
+        res_flag, op_name = check_norm_op(final_rmsnorm)
```
Comment (yfchen-byted) [inline xpu_graph/passes/patterns/targets/npu/fuse_add_rmsnorm.py:36]: 后续有待验证：增加fused_rms_norm op后，是否会影响这里判断

File: xpu_graph/passes/patterns/targets/npu/fuse_add_rmsnorm.py (hunk 14 above)
Comment (wxue1) [reply xpu_graph/passes/patterns/targets/npu/fuse_add_rmsnorm.py]: > 不要加，pattern manager会做

done

File: xpu_graph/passes/patterns/targets/npu/fuse_add_rmsnorm.py (hunk 15 above)
Comment (wxue1) [reply xpu_graph/passes/patterns/targets/npu/fuse_add_rmsnorm.py:36]: > 后续有待验证：增加fused_rms_norm op后，是否会影响这里判断

ok

File: tests/npu/test_silu_mul.py (hunk 13 above)
Comment (wxue1) [reply tests/npu/test_silu_mul.py:55]: > 留待后续修复：参考common目录test_fused_layernorm，通过caplog检查pass有没有生效

ok

**Review by yfchen-byted — COMMENTED** [review]
LGTM

**Review by yfchen-byted — APPROVED** [review]

**Comment by wxue1** [comment]
> 没有单元测试呢，补一下哈

补充了
